Feat/level 2

nakatomoyanyan opened the pull request:
⚪︎予想時間:8h
⚪︎実装時間:8h35m
⚪︎変更点の概要
・入力必須項目に入力がないとエラーになる
・タスクを削除できる
・完了と未完了を切り替えられる
・レスポンシブ対応
⚪︎実装した機能の指示内容
https://qiita.com/foolproof-htb/private/f612852ac603150d7b6f
⚪︎レイアウトの画像
・pc画面のレイアウト
<img width="1501" alt="TaskManagement_level2_pc" src="https://github.com/nakatomoyanyan/frontend_practice/assets/173334729/3934abc5-4e3e-419f-b434-e4fe1afbdc97">
・スマホ画面のレイアウト
<img width="604" alt="TaskManagement_level2_mobile" src="https://github.com/nakatomoyanyan/frontend_practice/assets/173334729/884f2c3c-3389-4df7-9391-01834cf320e0">
⚪︎動作のgif
![level2_4](https://github.com/nakatomoyanyan/frontend_practice/assets/173334729/ab7a54f3-cd68-4606-9816-0e0784e59711)

nakatomoyanyan committed:
フッターを画面下部に固定する

フッターを画面下部に固定しました。
タスクを追加した際にfooterのmargin-topが自動調整されるようにしました。

nakatomoyanyan committed:
タスクリストを横並びにして、画面の端までいったら折り返すようにする

タスクリストを横並びにしました。
画面の端までいったら折り返すようにしました。

nakatomoyanyan committed:
タスク間の間隔調整

タスク間の間隔を調整しました。

nakatomoyanyan committed:
フォームの高さを揃えて、各入力項目間に適度な間隔をあける

flexboxを用いてフォームの高さを揃えて、各入力項目間に10pxの間隔をあけました。

nakatomoyanyan committed:
幅 600px 以下の場合、タスク追加フォームとタスク一覧を縦並びにする

スマホ用の画面でタスク追加フォームとタスク一覧を縦並びにしました。
スマホ用画面で優先度のプルダウンの表示に不具合があります。

nakatomoyanyan committed:
フッターを画面下部に固定する(やり直し)

フッターを画面下部に固定しました。
以前のコミット時の方法では正しく固定できていなかったため、やり直しました。

nakatomoyanyan committed:
フッターの左側のずれを修正

フッターの左側のずれを修正しました。

nakatomoyanyan committed:
タスク名と期限を入力必須にする

タスク名と期限が一方でも未入力の場合はエラー文を表示させました。

nakatomoyanyan committed:
タスクの削除機能を実装する

タスクの削除ボタンをクリックすると、ダイアログ表示後にタスクが削除されるようにした。
タスク追加動作時にタスク削除ボタンも作成し、削除ボタンクリック時の動作を実装した。

nakatomoyanyan committed:
完了ボタンを押すと背景色が変わる

追加したタスク内に完了ボタンを追加した。また完了ボタンを押すと背景色が変更するようにした。背景色の変更は、javascriptでnewtaskに新しくクラスを追加することで実装した。

nakatomoyanyan committed:
タスクの完了ボタンをクリックする度にボタンの表示が切り替わる。

タスクの完了ボタンを押すと「完了にする」と「元に戻す」の表示が切り替わるようにしました。タスクの完了ボタンを押した際に「完了にする」がボタンに表示されている時は「元に戻す」に、それ以外の時は「完了する」に切り替えるようにif文を書きました。

nakatomoyanyan committed:
追加したタスクの「完了にする」ボタンを押すとタスクの名前に棒線が引かれる

タスクを完了した際にタスクの名前に棒線が引かれるようにしました。タスク完了ボタンを押すとnewtaskに「.is_complete」クラスが追加されて適応されるcssが切り替わります。

nakatomoyanyan committed:
タスク削除時のウィンドウにキャンセルも表示

タスク削除前にウィンドウで確認画面を表示。確認画面では「キャンセル」と「ok」を選択できるようにした。以前は「alert」を用いたが「comfirm」に変更した。これは「ok」だけではなく「キャンセル」も選べるようにするため。

nakatomoyanyan committed:
タスク削除時にウィンドウで「キャンセル」を選んだ際の動作取り消し

実装内容：タスク削除時にウィンドウで「キャンセル」を選ぶとタスクは削除されなくした。
実装方法：タスク削除時のウィンドウの選択内容を条件にif文を書いた。
実装理由：以前はウィンドウで「キャンセル」を選んでもタスクが削除されていた。

nakatomoyanyan committed:
JavaScript のコードを適当なサイズの関数に分割する

実装内容：フォームの初期化、削除ボタンの動作、完了ボタンの動作を関数でまとめた。

foolproof-htb commented on responsive.css:
この内容は別ファイルに分けなくても良さそうな気がしました！

foolproof-htb commented on task_add.js:
```diff
@@ -5,20 +5,59 @@ let taskImportance = document.querySelector('.task_importance'); //タスクの
+  if (window.confirm('本当に削除しますか？')) {
+    taskList.removeChild(newTask);
+  } else {
+  }
```
ここの else  は不要ですね 👀

foolproof-htb commented on task_add.js:
```diff
@@ -5,20 +5,59 @@ let taskImportance = document.querySelector('.task_importance'); //タスクの
+    taskSubmit.type = 'button';
+    initialization();
+  } else {
+    taskSubmit.type = 'submit';
```
すみません！！
私が `require` の挙動を勘違いしていたのですが、`type='button'` のときはエラー出力が動かないからここの記述があるんですね。
バックエンドへの送信がないのに「require を使う」という仕様にしている私の問題ですが、今回はHTMLのbuttonの方を `type='submit'` にしておいてもらえればと思います 🙇

foolproof-htb commented on task_add.js:
```diff
@@ -5,20 +5,59 @@ let taskImportance = document.querySelector('.task_importance'); //タスクの
-  taskList.appendChild(newTask);
-  //タスクの中身を追加
-  newTask.innerHTML = `
+  if (taskValue.value !== '' && taskDate.value !== '') {
```
if 文の中に処理全体を入れずに、以下のようにするとスマートかと思います。
```
if (taskValue.value === '' || taskDate.value === '') {
  taskSubmit.type = 'submit';
  return
}

// タスク追加処理
```

こういった書き方を早期リターンとかガード節とか言ったりします。

※ ちなみに、他のコメントで書いたような button タグの方の変更を行った場合は、`taskSubmit.type = 'submit';` も不要になります。

foolproof-htb commented on task_add.js:
```diff
@@ -5,20 +5,59 @@ let taskImportance = document.querySelector('.task_importance'); //タスクの
+  taskImportance.value = '高';
+}
+
+function removeIvent(newTask) {
```
細かいことですが、Ivent ではなく Event かもです！

foolproof-htb commented on task_add.js:
```diff
@@ -5,20 +5,59 @@ let taskImportance = document.querySelector('.task_importance'); //タスクの
+}
+
+function completeIvent(newTask, completeButton) {
+  newTask.classList.toggle('is_complete');
+  if (completeButton.innerText == '完了にする') {
+    completeButton.innerText = '元に戻す';
+  } else {
+    completeButton.innerText = '完了にする';
+  }
```
`toggle` は、指定した内容を付与したときに true、削除したときに false を返します。
これを利用して、
```
const isCompleteToggled = newTask.classList.toggle('is_complete');
completeButton.innerText = isCompleteToggled ? '元に戻す' : '完了にする';
```
みたいに三項演算子で書くと、重複も除けて簡潔に書けそうです！

iwaguma-manabi replied:
そもそも「元に戻す」状態のcompleteButtonは completeという名前でありながら、元に戻す仕事をすることになるというのが混乱の元になりそうな気がしました

```
ボタン・イベントは２つに分離して表示・非表示を切り替える
```

等で、責務と名称を合わせるのはどうでしょう？

foolproof-htb commented on task_add.js:
```diff
@@ -5,20 +5,59 @@ let taskImportance = document.querySelector('.task_importance'); //タスクの
+    newTask.appendChild(completeButton);
+    completeButton.addEventListener('click', () => {
+      completeIvent(newTask, completeButton);
+    });
```
ここまでの内容も全部 `completeIvent` 内に含めると、可読性とメンテナンスが上がるかと思います！
その際は「完了ボタンを追加する」という旨の関数名などに変更すると、より意味的に正しくなりそうです。
（remove も同様）

foolproof-htb commented on stylesheet.css:
```diff
@@ -20,17 +24,68 @@ header {
+  flex-direction: row;
+}
+
+.task_value {
+  margin-right: 10px;
+}
+
+.task_date {
+  margin-right: 10px;
+}
```
flex 下の要素は、`gap` で要素間の感覚を指定できます。
（厳密には横並びの感覚は column-gap だけど、今回は横一列しかないので gap でも可）

なので、flex を指定している form {} の中に gap を指定すると感覚を空けるのが楽になると思います！

foolproof-htb commented on stylesheet.css:
```diff
@@ -20,17 +24,68 @@ header {
+  padding-left: 10px;
+}
+
+.newtask.is_complete p:first-child {
+  text-decoration: line-through;
+}
```
タスクタイトルが必ずしも最初の要素とは限らないので、タスクタイトルにクラスを付与してそのクラスにスタイルを適用する方が柔軟性があっていいかと思います！

foolproof-htb reviewed (commented):
スタイル調整や機能追加など、全体的にいい感じだと思います 👍 
ただ、もっと別の書き方が出る部分などをいくつかコメントに書きましたので、ご確認いただければと思います！

nakatomoyanyan committed:
responsive.cssの削除

変更内容：responsive.cssの内容をstylesheet.cssにまとめた

nakatomoyanyan committed:
不要なelse文の削除

変更内容：削除ボタンの動作のif文の不要なelse文を削除しました。

nakatomoyanyan committed:
htmlの追加ボタンの属性をsubmitに変更＆if文の}忘れ解消

変更内容：htmlの追加ボタンの属性をsubmitに変更
　削除ボタンの挙動を記述するif文の}忘れの解消

nakatomoyanyan committed:
早期リターンの実装

変更内容：taskSbmit.addEventListenerに早期リターンを実装しました。実装に伴ってインデントを成形しました。

nakatomoyanyan committed:
関数名のスペルミス修正

修正内容：関数の名前に含まれていた「Ivent」を「Event」に変更しました。

nakatomoyanyan committed:
タスク完了ボタンとタスク未完了ボタンの機能分離＆タスク削除ボタンの追加を関数化

変更内容：タスク完了機能とタスク未完了機能を別々の関数で記述。合わせてタスク未完了ボタンのcssを新たに記述。また、タスク削除の機能を関数でまとめて記述。

nakatomoyanyan committed:
タスクの中身の間隔をgapで指定

変更内容：タスクの中身の間隔をgapで指定するように変更。

nakatomoyanyan committed:
タスク完了ボタン＆タスク未完了ボタン＆タスク削除ボタンの追加方法の変更＆初期化の記述の変更

変更内容：タスク追加、完了、削除ボタンをタスク追加時に作成して、ボタンのクリックに応じて表示させたり非表示にしたりする。
resetメソッドを使ってフォームを初期化するように変更。

nakatomoyanyan committed:
タスクタイトルにクラス付与

変更内容：タスクタイトルのタグにクラス付与。合わせてcssの記述を変更。
